CEXT-3214: Docs for consuming events (#219)
AdobeDocs/commerce-extensibility: src/pages/events/consume-events.md

      event_payload.extension_attributes = content.extension_attributes
    }

    // Send event data to a third-party ERP system
    sendOrderToErpSystem(event_payload)
      
    return {
      statusCode: 200,
      body: event_payload
    }
  } catch (error) {
    // log any server errors
    logger.error(error)
    // return with 500
    return errorResponse(500, 'server error', logger)
  }
}
  
exports.main = main
```

The `main` actions first accesses the payload for a received `observer.sales_order_save_after` event:
```js
event_payload = params.data.value
```

The event payload for this event may not contain the saved order's extension attributes. In that case, the extension attributes can be fetched for the specific order using a Commerce API call:
```js
  const oauth = getCommerceOauthClient(
    {
      url: params.COMMERCE_BASE_URL,
      consumerKey: params.COMMERCE_CONSUMER_KEY,
      consumerSecret: params.COMMERCE_CONSUMER_SECRET,
      accessToken: params.COMMERCE_ACCESS_TOKEN,
      accessTokenSecret: params.COMMERCE_ACCESS_TOKEN_SECRET
    },
    logger
  )
  const content = await oauth.get('orders/' + event_payload.entity_id)
```

> In this example, functions from the `oauth1a` module, as defined in the [adobe-commerce-samples repo](https://github.com/adobe/adobe-commerce-samples/blob/main/sample-extension/actions/oauth1a.js), are used. 

Author: TODO: this may move as part of https://github.com/adobe/adobe-commerce-samples/pull/9

  - Extensibility
---

Events sent from Adobe Commerce to Adobe I/O events can be consumed in a few different ways. The following options for consuming events can be configured when adding events to your App Builder project and creating an event registration:

* [Using the Journaling API](#using-the-journaling-api)
* [Using a Webhook URL](#using-a-webhook-url)
* [Using a Runtime Action](#using-a-runtime-action)
* [Using Amazon EventBridge](#using-amazon-eventbridge)

## Using the Journaling API

When an Adobe I/O event registration is created, the subscribed events will by default be added to an ordered list, referred to as the journal. These events can be consumed using a journaling endpoint URL that is unique to the registration. For more information on reading events from the journal, see the [Introduction to Journaling](https://developer.adobe.com/events/docs/guides/journaling_intro/).

Events can also be consumed from the Journaling API using the [Adobe I/O Events SDK](https://github.com/adobe/aio-lib-events). See [Subscribe to Events Using Journaling](https://developer.adobe.com/events/docs/guides/sdk/sdk_journaling/) for details and sample code.

## Using a Webhook URL

> When creating or editing an event registration, a webhook URL can be registered. Doing so will cause subscribed Commerce events sent to Adobe I/O Events to be forwarded to the specified webhook URL. See [Introduction to Adobe I/O Events Webhooks](https://developer.adobe.com/events/docs/guides/) for more details.

Reviewer: Can we add a bit more context like this "This option is useful for use cases that do not require App Builder to handle transformations to the event structure or additional API calls. This method of consuming events allows for easily forwarding events from Adobe Commerce to a 3rd party system via App Builder. Note that the destination system must have functionality to support a webhook URL endpoint in order to use this option." 

I think a screenshot of the webhook event registration UI would be useful here as well

Reviewer: @sarahxxu do you think it would be helpful to include a note like "Be careful not to confuse this method of consuming events with [Runtime Actions as a Webhook] (https://developer.adobe.com/events/docs/guides/runtime_webhooks/#benefits-of-using-runtime-action-as-webhook). Runtime Actions as a Webhook is a functionality for creating a URL for App Builder to receive events from a 3rd party system. 

Personally I got very confused by this initially

Author: I've added this statement as an inline warning within this section for now

## Using a Runtime Action

An [Adobe I/O Runtime Action](https://developer.adobe.com/runtime/docs/guides/overview/entities/#actions) can be set up to receive Commerce events in an event registration. Actions can be created from JavaScript functions, as described in [Creating Actions](https://developer.adobe.com/runtime/docs/guides/using/creating_actions/). Within an action, business logic can be executed based on the received event payload, API calls can be made back to Adobe Commerce to update data or access additional information, and data can be sent to another system.

See [Runtime Action with a Callback to Commerce](./consume-events-examples/runtime-action-commerce-callback.md) for an example of how a runtime action can consume an event and make an API call back to Commerce to get more information.

> 

Reviewer: Maybe we can add a reference to the starter kit here, saying something like "See more examples of end-to-end integrations between Commerce and 3rd party systems like ERPs in our [Starter Kit] (https://developer.adobe.com/commerce/extensibility/starter-kit/send-data/) documentation. 

WDYT @jhadobe

Reviewer 2: @naomirobichaud i think it makes sense to reference the starter kit docs. we should also likely reference the code samples repo. in fact, some of these code samples that @ryanjcohen provided might be better suited there: https://github.com/adobe/adobe-commerce-samples

Author: I thought about the adobe-commerce-samples repo too. The example in `runtime-action-commerce-callback.md`  isn't fully functional though, just because of the `await sendOrderToErpSystem(event_payload)` line. `sendOrderToErpSystem` and the `erp` module don't actually exist. I just put them in to illustrate that the event payload could be sent somewhere after data modification. I was thinking that code examples in adobe-commerce-samples are more complete/ready for creating app builder actions for experimenting. What are your thoughts on this @naomirobichaud?

## Using a Webhook URL

Consuming an event using a webhook URL is useful for cases that do not require App Builder to handle transformations to the event structure or additional API calls, which simplifies forwarding events from Adobe Commerce to a 3rd-party system using App Builder. The destination system must have an endpoint that can receive webhook requests to use this option.

You can register the webhook URL when creating or editing an event registration. If you are setting up the registration through the developer console UI, this option will appear in the final configuration step.

![Webhook registration](../_images/events/register-webhook.png)

See [Introduction to Adobe I/O Events Webhooks](https://developer.adobe.com/events/docs/guides/) for more details.

<InlineAlert variant="warning" slots="text"/>

Be careful not to confuse this method of consuming events with [Runtime Actions as a Webhook](https://developer.adobe.com/events/docs/guides/runtime_webhooks/#benefits-of-using-runtime-action-as-webhook). Runtime Actions as a Webhook is a functionality for creating a URL for App Builder to receive events from a 3rd party system.

> ## Using a Runtime Action
> 
> 

Reviewer 2: ```suggestion
## Using a Runtime Action

```

---
title: Consume Events
description: Learn how to consume events sent from Adobe Commerce to Adobe I/O Events.
keywords:
  - Events
  - Extensibility
---

> Events sent from Adobe Commerce to Adobe I/O events can be consumed in a few different ways. The following options for consuming events are available when adding events to your App Builder project and creating an event registration:

Reviewer 2: ```suggestion
You can consume events sent from Adobe Commerce to Adobe I/O events in several ways. The following options for consuming events are available when adding events to your App Builder project and creating an event registration:
```

* [Using the Journaling API](#using-the-journaling-api) (enabled by default)
* [Using a Webhook URL](#using-a-webhook-url)
* [Using a Runtime Action](#using-a-runtime-action)
* [Using Amazon EventBridge](#using-amazon-eventbridge)

## Using the Journaling API

> When an Adobe I/O event registration is created, the subscribed events will by default be added to an ordered list, referred to as the journal. These events can be consumed using a journaling endpoint URL that is unique to the registration. For more information on reading events from the journal, see the [Introduction to Journaling](https://developer.adobe.com/events/docs/guides/journaling_intro/).

Reviewer 2: ```suggestion
When you create an Adobe I/O event registration, the subscribed events added to an ordered list, referred to as the journal, by default. You can consume these events using a journaling endpoint URL that is unique to the registration. For more information on reading events from the journal, see the [Introduction to Journaling](https://developer.adobe.com/events/docs/guides/journaling_intro/).
```

> Events can also be consumed from the Journaling API using the [Adobe I/O Events SDK](https://github.com/adobe/aio-lib-events). See [Subscribe to Events Using Journaling](https://developer.adobe.com/events/docs/guides/sdk/sdk_journaling/) for details and sample code.

Reviewer 2: ```suggestion
You can also consume events from the Journaling API using the [Adobe I/O Events SDK](https://github.com/adobe/aio-lib-events). See [Subscribe to Events Using Journaling](https://developer.adobe.com/events/docs/guides/sdk/sdk_journaling/) for details and sample code.
```

## Using a Webhook URL

> Consuming an event using a webhook URL is useful for cases that do not require App Builder to handle transformations to the event structure or additional API calls. This method of consuming events allows for easily forwarding events from Adobe Commerce to a 3rd party system via App Builder. Note that the destination system must have an endpoint that can receive webhook requests in order to use this option.

Reviewer 2: ```suggestion
Consuming an event using a webhook URL is useful for cases that do not require App Builder to handle transformations to the event structure or additional API calls, which simplifies forwarding events from Adobe Commerce to a 3rd-party system using App Builder. The destination system must have an endpoint that can receive webhook requests to use this option.
```

> The webhook URL can be registered when creating or editing an event registration. If the registration is being set up through the developer console UI, this option will appear in the final configuration step.

Reviewer 2: ```suggestion
You can register the webhook URL when creating or editing an event registration. If you are setting up the registration through the developer console UI, this option will appear in the final configuration step.
```

![Webhook registration](../_images/events/register-webhook.png)

See [Introduction to Adobe I/O Events Webhooks](https://developer.adobe.com/events/docs/guides/) for more details.

<InlineAlert variant="warning" slots="text"/>

Be careful not to confuse this method of consuming events with [Runtime Actions as a Webhook](https://developer.adobe.com/events/docs/guides/runtime_webhooks/#benefits-of-using-runtime-action-as-webhook). Runtime Actions as a Webhook is a functionality for creating a URL for App Builder to receive events from a 3rd party system.

## Using a Runtime Action

> An [Adobe I/O Runtime Action](https://developer.adobe.com/runtime/docs/guides/overview/entities/#actions) can be set up to receive Commerce events in an event registration. Actions can be created from JavaScript functions, as described in [Creating Actions](https://developer.adobe.com/runtime/docs/guides/using/creating_actions/). Within an action, business logic can be executed based on the received event payload, API calls can be made back to Adobe Commerce to update data or access additional information, and data can be sent to another system.

Reviewer 2: ```suggestion
```
An [Adobe I/O Runtime Action](https://developer.adobe.com/runtime/docs/guides/overview/entities/#actions) allows you to receive Commerce events in an event registration. You can create actions from JavaScript functions, as described in [Creating Actions](https://developer.adobe.com/runtime/docs/guides/using/creating_actions/). Within an action, you can execute business logic based on the received event payload, make API calls back to Adobe Commerce to update data, or access additional information, and send data to another system.

Author: @jhadobe Just double checking, was that comment meant to be included as text in the `Using a Runtime Action` section? I think the previous runtime action description was removed and not replaced

Reviewer 2: ah, yes, i missed it. i'll update in a subsequent PR.

See [Runtime Action with a Callback to Commerce](./consume-events-examples/runtime-action-commerce-callback.md) for an example of how a runtime action can consume an event and make an API call back to Commerce to get more information.

See more examples of end-to-end integrations between Commerce and 3rd party systems like ERPs that use runtime actions in our [Starter Kit](https://developer.adobe.com/commerce/extensibility/starter-kit/send-data/) documentation.

## Using Amazon EventBridge

> An Adobe I/O event registration can be configured to forward received Commerce events to Amazon EventBridge. See [Adobe I/O Events and Amazon EventBridge Integration](https://developer.adobe.com/events/docs/guides/amazon_eventbridge/) for more details.

Reviewer 2: ```suggestion
You can configure an Adobe I/O event registration to forward received Commerce events to Amazon EventBridge. See [Adobe I/O Events and Amazon EventBridge Integration](https://developer.adobe.com/events/docs/guides/amazon_eventbridge/) for more details.
```
